Impossible to install @angular-devkit/build-angular ~0.1102.13 on npm7

<!--🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅

Oh hi there! 😄

To expedite issue processing please search open and closed issues before submitting a new one.
Existing issues often contain information about workarounds, resolution, or progress updates.

🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅-->

# 🐞 Bug report

### Command (mark with an `x`)

<!-- Can you pin-point the command or commands that are effected by this bug? -->
<!-- ✍️edit: -->

- [ ] new
- [ ] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] extract-i18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc

### Is this a regression?

<!-- Did this behavior use to work in the previous version? -->
<!-- ✍️--> Yes, the previous version in which this bug was not present was: ....

### Description
It's impossible to install @angular-devkit/build-angular ~0.1102.13 if using npm7
It uses postcss 8.2.14, but also cssnano 5.0.2, which uses cssnano-preset-default@5.1.0, which has a peer dependency to postcss@"^8.2.15", so npm 7 fails to install because of peer dependency mismatch
<!-- ✍️--> A clear and concise description of the problem...

## 🔬 Minimal Reproduction

```
mkdir justatest
cd justatest
npm init 
npm install @angular-devkit/build-angular@~0.1102.13
```
<!--
Simple steps to reproduce this bug.

Please include: commands run (including args), packages added, related code changes.

If reproduction steps are not enough for reproduction of your issue, please create a minimal GitHub repository with the reproduction of the issue.
A good way to make a minimal reproduction is to create a new app via `ng new repro-app` and add the minimum possible code to show the problem.
Share the link to the repo below along with step-by-step instructions to reproduce the problem, as well as expected and actual behavior.

Issues that don't have enough info and can't be reproduced will be closed.

You can read more about issue submission guidelines here: https://github.com/angular/angular-cli/blob/master/CONTRIBUTING.md#-submitting-an-issue
-->

## 🔥 Exception or Error

<pre><code>
<!-- If the issue is accompanied by an exception or an error, please share it below: -->
<!-- ✍️-->
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! Found: postcss@8.2.14
npm ERR! node_modules/postcss
npm ERR!   postcss@"8.2.14" from @angular-devkit/build-angular@0.1102.13
npm ERR!   node_modules/@angular-devkit/build-angular
npm ERR!     @angular-devkit/build-angular@"~0.1102.13" from the root project
npm ERR!   peer postcss@"^8.1.0" from autoprefixer@10.2.4
npm ERR!   node_modules/autoprefixer
npm ERR!     autoprefixer@"10.2.4" from @angular-devkit/build-angular@0.1102.13
npm ERR!     node_modules/@angular-devkit/build-angular
npm ERR!       @angular-devkit/build-angular@"~0.1102.13" from the root project
npm ERR!   9 more (css-loader, cssnano, postcss-import, postcss-loader, ...)
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer postcss@"^8.2.15" from cssnano-preset-default@5.1.0
npm ERR! node_modules/cssnano/node_modules/cssnano-preset-default
npm ERR!   cssnano-preset-default@"^5.0.1" from cssnano@5.0.2
npm ERR!   node_modules/cssnano
npm ERR!     cssnano@"5.0.2" from @angular-devkit/build-angular@0.1102.13
npm ERR!     node_modules/@angular-devkit/build-angular
npm ERR!       @angular-devkit/build-angular@"~0.1102.13" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
</code></pre>

## 🌍 Your Environment


no need to use ng command, it fails if you just try to install the package
<pre><code>
<!-- run `ng version` and paste output below -->
<!-- ✍️-->
npm 7
</code></pre>

**Anything else relevant?**

possible solution is to update the postcss dependency to 8.2.15, our use ~8.2.15 so it doesn't break again anytime soon
<!-- ✍️Is this a browser specific issue? If so, please specify the browser and version. -->

<!-- ✍️Do any of these matter: operating system, IDE, package manager, HTTP server, ...? If so, please mention it below. -->

--- a/package.json
+++ b/package.json
@@ -192,7 +192,7 @@
     "pidusage": "^2.0.17",
     "pnp-webpack-plugin": "1.6.4",
     "popper.js": "^1.14.1",
-    "postcss": "8.2.14",
+    "postcss": "8.2.15",
     "postcss-import": "14.0.0",
     "postcss-loader": "4.2.0",
     "prettier": "^2.0.0",

--- a/packages/angular_devkit/build_angular/package.json
+++ b/packages/angular_devkit/build_angular/package.json
@@ -50,7 +50,7 @@
     "ora": "5.3.0",
     "parse5-html-rewriting-stream": "6.0.1",
     "pnp-webpack-plugin": "1.6.4",
-    "postcss": "8.2.14",
+    "postcss": "8.2.15",
     "postcss-import": "14.0.0",
     "postcss-loader": "4.2.0",
     "raw-loader": "4.0.2",

--- a/tests/legacy-cli/e2e/setup/010-local-publish.ts
+++ b/tests/legacy-cli/e2e/setup/010-local-publish.ts
@@ -1,11 +1,10 @@
-import { prerelease } from 'semver';
-import { packages } from '../../../../lib/packages';
 import { getGlobalVariable } from '../utils/env';
 import { npm } from '../utils/process';
+import { isPrereleaseCli } from '../utils/project';
 
-export default async function() {
+export default async function () {
   const testRegistry = getGlobalVariable('package-registry');
-  const publishArgs = [
+  await npm(
     'run',
     'admin',
     '--',
@@ -13,14 +12,7 @@ export default async function() {
     '--no-versionCheck',
     '--no-branchCheck',
     `--registry=${testRegistry}`,
-  ];
-
-  const pre = prerelease(packages['@angular/cli'].version);
-  if (pre && pre.length > 0) {
-    publishArgs.push('--tag', 'next');
-  } else {
-    publishArgs.push('--tag', 'latest');
-  }
-
-  await npm(...publishArgs);
+    '--tag',
+    isPrereleaseCli() ? 'next' : 'latest',
+  );
 }

--- a/tests/legacy-cli/e2e/tests/update/update-8.ts
+++ b/tests/legacy-cli/e2e/tests/update/update-8.ts
@@ -1,23 +1,43 @@
 import { createProjectFromAsset } from '../../utils/assets';
 import { expectFileMatchToExist } from '../../utils/fs';
-import { installWorkspacePackages } from '../../utils/packages';
-import { ng, noSilentNg, silentNpm } from '../../utils/process';
-import { isPrereleaseCli, useBuiltPackages, useCIChrome, useCIDefaults } from '../../utils/project';
+import { installPackage, installWorkspacePackages, setRegistry } from '../../utils/packages';
+import { ng, noSilentNg } from '../../utils/process';
+import { isPrereleaseCli, useCIChrome, useCIDefaults } from '../../utils/project';
 
-export default async function() {
-  await createProjectFromAsset('8.0-project');
-  await ng('update', '@angular/cli', '--migrate-only', '--from=8');
+export default async function () {
+  // We need to use the public registry because in the local NPM server we don't have
+  // older versions @angular/cli packages which would cause `npm install` during `ng update` to fail.
+  try {
+    await createProjectFromAsset('8.0-project', true, true);
 
-  // Use the packages we are building in this commit, and CI Chrome.
-  await useBuiltPackages();
+    await setRegistry(false);
+    await installWorkspacePackages();
+
+    // Update Angular to 9
+    await installPackage('@angular/cli@8');
+    const { stdout } = await ng('update', '@angular/cli@9.x', '@angular/core@9.x');
+    if (!stdout.includes("Executing migrations of package '@angular/cli'")) {
+      throw new Error('Update did not execute migrations. OUTPUT: \n' + stdout);
+    }
+
+    // Update Angular to 10
+    await ng('update', '@angular/cli@10', '@angular/core@10');
+  } finally {
+    await setRegistry(true);
+  }
+
+  // Update Angular current build
+  const extraUpdateArgs = isPrereleaseCli() ? ['--next', '--force'] : ['--force'];
+  // For the latest/next release we purposely don't add `@angular/core`.
+  // This is due to our bumping strategy, which causes a period were `@angular/cli@latest` (v12.0.0) `@angular/core@latest` (v11.2.x)
+  // are of different major/minor version on the local NPM server. This causes `ng update` to fail.
+  // NB: `ng update @angula/cli` will still cause `@angular/core` packages to be updated.
+  await ng('update', '@angular/cli', ...extraUpdateArgs);
+
+  // Setup testing to use CI Chrome.
   await useCIChrome('./');
   await useCIChrome('./e2e/');
   await useCIDefaults('eight-project');
-  await installWorkspacePackages();
-
-  // Update Angular.
-  const extraUpdateArgs = await isPrereleaseCli() ? ['--next', '--force'] : [];
-  await ng('update', '@angular/core', ...extraUpdateArgs);
 
   // Run CLI commands.
   await ng('generate', 'component', 'my-comp');

--- a/tests/legacy-cli/e2e/utils/packages.ts
+++ b/tests/legacy-cli/e2e/utils/packages.ts
@@ -1,5 +1,6 @@
 import { getGlobalVariable } from './env';
-import { ProcessOutput, silentNpm, silentYarn } from './process';
+import { writeFile } from './fs';
+import { npm, ProcessOutput, silentNpm, silentYarn } from './process';
 
 export function getActivePackageManager(): 'npm' | 'yarn' {
   const value = getGlobalVariable('package-manager');
@@ -37,5 +38,23 @@ export async function uninstallPackage(name: string): Promise<ProcessOutput> {
       return silentNpm('uninstall', name);
     case 'yarn':
       return silentYarn('remove', name);
+  }
+}
+
+export async function setRegistry(useTestRegistry: boolean): Promise<void> {
+  const url = useTestRegistry
+    ? getGlobalVariable('package-registry')
+    : 'https://registry.npmjs.org';
+
+  const isCI = getGlobalVariable('ci');
+
+  // Ensure local test registry is used when outside a project
+  if (isCI) {
+    // Safe to set a user configuration on CI
+    await npm('config', 'set', 'registry', url);
+  } else {
+    // Yarn does not use the environment variable so an .npmrc file is also required
+    await writeFile('.npmrc', `registry=${url}`);
+    process.env['NPM_CONFIG_REGISTRY'] = url;
   }
 }

--- a/tests/legacy-cli/e2e/utils/project.ts
+++ b/tests/legacy-cli/e2e/utils/project.ts
@@ -7,9 +7,6 @@ import { gitCommit } from './git';
 import { installWorkspacePackages } from './packages';
 import { execAndWaitForOutputToMatch, git, ng, npm } from './process';
 
-const tsConfigPath = 'tsconfig.json';
-
-
 export function updateJsonFile(filePath: string, fn: (json: any) => any | void) {
   return readFile(filePath)
     .then(tsConfigJson => {
@@ -23,7 +20,7 @@ export function updateJsonFile(filePath: string, fn: (json: any) => any | void) 
 
 
 export function updateTsConfig(fn: (json: any) => any | void) {
-  return updateJsonFile(tsConfigPath, fn);
+  return updateJsonFile('tsconfig.json', fn);
 }
 
 
@@ -293,9 +290,8 @@ export function useCIChrome(projectDir: string) {
     });
 }
 
-export async function isPrereleaseCli() {
-  const angularCliPkgJson = JSON.parse(await readFile('node_modules/@angular/cli/package.json'));
-  const pre = prerelease(angularCliPkgJson.version);
+export function isPrereleaseCli(): boolean {
+  const pre = prerelease(packages['@angular/cli'].version);
 
   return pre && pre.length > 0;
 }

